src/mesh/region_set.cc

```diff
@@ -0,0 +1,385 @@
+		try {
+			reg = mesh->region_db().find_id(region_id);
+		} catch(RegionDB::ExcUniqueRegionId &e) {
+			e << rec.ei_address();
+		}
+		if ( reg.is_valid() ) {
+			region_db_.rename_region(reg, region_label);
+		} else {
+			stringstream ss;
+			ss << "declare at address \'" << rec.address_string() << "\'";
+			region_db_.add_region(region_id, region_label, RegionDB::undefined_dim, ss.str() );
+		}
+	}
+}
+
+
+
+const IT::Record & RegionSetFromId::get_region_input_type()
+{
+    return IT::Record("From_Id", "Region declared by id and name.\n"
+    					"Allows to create new region with given id and label\n"
+    					"or specify existing region by id which will be renamed.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"Label (name) of the region. Has to be unique in one mesh.\n")
+		.declare_key("id", IT::Integer(0), IT::Default::obligatory(),
+				"The ID of the region to which you assign label.")
+		.close();
+}
+
+
+
+const int RegionSetFromId::registrar =
+		Input::register_class< RegionSetFromId, const Input::Record &, Mesh * >("From_Id") +
+		RegionSetFromId::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetFromLabel
+ */
+
+RegionSetFromLabel::RegionSetFromLabel(const Input::Record &rec, Mesh *mesh)
+: RegionSetBase(mesh)
+{
+	string new_name = rec.val<string>("name");
+	string mesh_label = rec.val<string>("mesh_label");
+
+	Region reg = mesh->region_db().find_label(mesh_label);
+	if ( reg.is_valid() ) {
+		region_db_.rename_region(reg, new_name);
+	} else {
+		xprintf(Warn, "Unknown region in mesh with label '%s'\n", mesh_label.c_str());
+	}
+}
+
+
+
+const IT::Record & RegionSetFromLabel::get_region_input_type()
+{
+    return IT::Record("From_Label", "Allows to rename existing region specified by mesh_label.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"New label (name) of the region. Has to be unique in one mesh.")
+		.declare_key("mesh_label", IT::String(), IT::Default::obligatory(),
+				"The mesh_label is e.g. physical volume name in GMSH format.")
+		.close();
+}
+
+
+
+const int RegionSetFromLabel::registrar =
+		Input::register_class< RegionSetFromLabel, const Input::Record &, Mesh * >("From_Label") +
+		RegionSetFromLabel::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetFromElements
+ */
+
+RegionSetFromElements::RegionSetFromElements(const Input::Record &rec, Mesh *mesh)
+: RegionSetBase(mesh)
+{
+	unsigned int region_id;
+	string region_label = rec.val<string>("name");
+	Region reg = mesh->region_db().find_label(region_label);
+
+	Input::Iterator<unsigned int> it = rec.find<unsigned int>("id");
+
+	if ( reg.is_valid() ) { // region with the specified label exists
+		region_id = reg.id();
```

Comment:
Rather throw always if the label already exists and change the exception message (region of same label already exists), current message is difficult to understand. Also check if RegionDB do not declare a similar exception. Could be reused.